source/agora/common/MerkleTree.d

```diff
@@ -0,0 +1,233 @@
+
+    /// The right branch's ptr of the tree.
+    public MerkleNode* right;
+
+    /// The transaction's ptr
+    public Transaction* tx;
+
+    /// The hash of this `MerkleNode`
+    public Hash hash;
+
+    /***************************************************************************
+
+        Create
+
+        Params:
+            left = The left branch's ptr of the tree.
+            right = The right branch's ptr of the tree.
+            tx = The transaction's ptr
+
+    ***************************************************************************/
+
+    /// Ctor
+    this (MerkleNode* left, MerkleNode* right, Transaction* tx = null)
+    {
+        this.left = left;
+        this.right = right;
+        this.tx = tx;
+        this.hash = this.getHash();
+    }
+
+    /***************************************************************************
+
+        Calculate the hash value of the `MerkleNode`
+
+        Return:
+            The `Hash` representing this instance
+
+    ***************************************************************************/
+
+    public Hash getHash ()
+        nothrow @nogc @trusted
+    {
+        if ((this.left is null) && (this.right is null))
+        {
+            if (this.tx !is null)
+                return hashFull(*this.tx);
+            else
+                return Hash.init;
+        }
+        else if ((this.left !is null) && (this.right !is null))
+        {
+            return mergeHash(this.left.hash, this.right.hash);
+        }
+        else
+            return Hash.init;
+    }
+}
+
+
+/*******************************************************************************
+
+    Found nodes with specific hash on child nodes.
+
+    Params:
+        parent = parent node.
+        hash = looking hash.
+
+    Returns:
+        Return node if found the hash. Return null otherwise.
+
+*******************************************************************************/
+
+public MerkleNode* searchNode (MerkleNode* parent, Hash hash)
+{
+    if (parent is null) return null;
+
+    if (parent.hash == hash) return parent;
+
+    MerkleNode* res;
+    res = searchNode(parent.right, hash);
+    if (res != null) return res;
+
+    res = searchNode(parent.left, hash);
+    if (res != null) return res;
+
+    return null;
+}
+
+/// Merge two hash
+public Hash mergeHash (Hash h1, Hash h2)
+        nothrow @nogc @trusted
+{
+    static struct MergeHash
+    {
+        Hash left;
+        Hash right;
+
+        public void computeHash (scope HashDg dg) const nothrow @safe @nogc
+        {
+            if (this.left < this.right)
+            {
+                hashPart(this.left, dg);
+                hashPart(this.right, dg);
+            }
+            else
+            {
+                hashPart(this.right, dg);
+                hashPart(this.left, dg);
+            }
+        }
+    }
+
+    return hashFull(MergeHash(h1, h2));
+}
+
+/// If there is no transaction, the hash value will be the initial value.
```

Comment:
I don't think we should be able to create a tree without transaction. There is no use case for it.

Reply:
Ok